### PR TITLE
fix #594 - deadlock in unlock() when called from Python

### PR DESCRIPTION
**Pull request by pinkavaj**
See comments in individula patches for details.

I give it some testing but it should be inspected more thoroughly.

Side note:
This fix allows some more intriguing code to be writen and thus some more bugs probably appear.
I personaly fall into "sched: <block fft_filter_ccc (19)> is requesting more input data".

### CODE REVIEW

File: gnuradio-runtime/lib/top_block_impl.cc [hunk 1]
```diff
@@ -129,6 +132,21 @@ namespace gr {
   void
   top_block_impl::wait()
   {
+    do {
```
Comment (jmcorgan) [inline gnuradio-runtime/lib/top_block_impl.cc]: Can you explain this do {} while(false); pattern?

File: gr-blocks/python/blocks/qa_hier_block2.py [hunk 2]
```diff
@@ -426,5 +427,20 @@ def test_033c_get_affinity(self):
...
         self.assertEquals((0,), procs)
 
+    def test_lock_unlock(self):
+        hblock = gr.top_block("test_block")
+        src = blocks.null_source(gr.sizeof_float)
+        throttle = blocks.throttle(gr.sizeof_float, 32000)
+        hier = multiply_const_ff(0.5)
+        sink = blocks.null_sink(gr.sizeof_float)
+        hblock.connect(src, throttle, hier, sink)
+        hblock.set_processor_affinity([0,])
```
Comment (jmcorgan) [inline gr-blocks/python/blocks/qa_hier_block2.py]: Why is this needed?

**Comment by jmcorgan** [comment]
Please rebase this off the maint branch.

File: gr-blocks/python/blocks/qa_hier_block2.py (hunk 2 above)
Comment (pinkavaj) [inline gr-blocks/python/blocks/qa_hier_block2.py]: this is a mistake, wrong cherry-pick, it hould have been test for lock()/unlock() sequence, will push -f fix

File: gnuradio-runtime/lib/top_block_impl.cc (hunk 1 above)
Comment (pinkavaj) [inline gnuradio-runtime/lib/top_block_impl.cc]: It waits for jobs to complete. Jobs complete after call of stop(). The checks whatever the block is locked, if not then exit, if is locked, wait until is unlocked and do both checks again. The point is to prevent exiting from wait() whet the block is locked but stoped. Until now this behaviour was simulated trought consectuive calls of stop() wait() start() wich does not given to other calls of wait() enough time to rally exit.

Comment (jmcorgan) [inline gnuradio-runtime/lib/top_block_impl.cc]: What I was really asking was the need for do {} while(false) as an outer envelope for the rest of the code.

**Comment by jmcorgan** [comment]
Please resubmit this as a new PR, off the maint branch.

File: gnuradio-runtime/lib/top_block_impl.cc (hunk 1 above)
Comment (pinkavaj) [inline gnuradio-runtime/lib/top_block_impl.cc]: Just to add proper boundaries for lock. The d_lock_count  must be examined under lock, but wait_for_jobs() must be in unlocked section of the code.

This is just loop, but is loopend through the 'continue' , not trought the condition on while().

I still have the feeling I did not get what You want to now.

Comment (jmcorgan) [inline gnuradio-runtime/lib/top_block_impl.cc]: It's unneeded, and optimized away by the compiler anyway.  The local scope for the scoped_lock is provided by the opening and closing braces on lines 137 and 143.

Comment (pinkavaj) [inline gnuradio-runtime/lib/top_block_impl.cc]: You are right, the test condition is always checked even if continue is used. So this is wrong, I will rewrite and pus to the new PR.
